### PR TITLE
Transactional email

### PR DESCRIPTION
**Pull request by sdelamo**
You can generate docs with `./gradlew docs`

### CODE REVIEW

File: email/src/main/java/io/micronaut/email/TransactionalEmail.java [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+     * @param from Sender of the Email
+     * @param to Email's recipient
+     * @param subject Email's subject
+     */
+    public TransactionalEmail(@NonNull Contact from,
+                              @NonNull List<Contact> to,
+                              @NonNull String subject) {
+        this(from, to, subject, null, null, null, null, null);
+    }
+
```
Comment (ilopmar) [inline email/src/main/java/io/micronaut/email/TransactionalEmail.java:68]: If we want to make users to use the `Builder` class maybe it's better to have just one private constructor.

File: email-sendgrid/src/test/groovy/io/micronaut/email/TransactionalEmailSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+import io.micronaut.core.beans.BeanIntrospection
+import io.micronaut.test.extensions.spock.annotation.MicronautTest
+import jakarta.inject.Inject
+import spock.lang.PendingFeature
+import spock.lang.Specification
+import javax.validation.Validator
+
+@MicronautTest(startApplication = false)
+class TransactionalEmailSpec extends Specification {
```
Comment (ilopmar) [inline email-sendgrid/src/test/groovy/io/micronaut/email/TransactionalEmailSpec.groovy:11]: This class should be in `email` module

File: email/src/main/java/io/micronaut/email/TransactionalEmail.java [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    @NonNull
+    @NotNull
+    @Valid
+    private final Sender sender;
+
+    @NonNull
+    @NotNull
+    @Valid
+    private final Recipient recipient;
```
Comment (ilopmar) [inline email/src/main/java/io/micronaut/email/TransactionalEmail.java:45]: It feels strange to use here the names `sender` and `recipient` and then use `from` and `to`.

File: email-mailjet/build.gradle [hunk 4]
```diff
@@ -0,0 +1,13 @@
+plugins {
+    id("io.micronaut.build.internal.module")
+}
+
+group = "io.micronaut.email"
+
+dependencies {
+    annotationProcessor("io.micronaut:micronaut-validation")
+    implementation("io.micronaut:micronaut-validation")
+    implementation("com.mailjet:mailjet-client:5.2.0")
```
Comment (ilopmar) [inline email-mailjet/build.gradle:10]: The version should be in `gradle.properties` so we can also have it in core's BOM. This comment applies to all modules.

File: email-postmark/build.gradle [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    id("io.micronaut.build.internal.module")
+}
+
+group = "io.micronaut.email"
+
+dependencies {
+    annotationProcessor("io.micronaut:micronaut-validation")
+    implementation("com.wildbit.java:postmark:1.7.5")
+    implementation("io.micronaut:micronaut-validation")
+    implementation("com.sendgrid:sendgrid-java:4.8.0")
```
Comment (ilopmar) [inline email-postmark/build.gradle:11]: Remove sendgrid dependency from here

File: email-mailjet/src/test/groovy/io/micronaut/email/mailjet/MailjetEmailSenderSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import spock.util.concurrent.PollingConditions
+
+@MicronautTest(startApplication = false)
+class MailjetEmailSenderSpec extends Specification {
+
+    @Inject
+    EmailCourier emailCourier
+
+    @Requires({env["MAILJET_API_KEY"] && env["MAILJET_API_SECRET"] && env["GMAIL_USERNAME"] && env["GMAIL_PASSWORD"]})
+    void "Functional test of SendGrid integration"() {
```
Comment (ilopmar) [inline email-mailjet/src/test/groovy/io/micronaut/email/mailjet/MailjetEmailSenderSpec.groovy:19]: Typo `s/sendgrid/mailjet`

**Review by ilopmar — CHANGES_REQUESTED** [review]
It looks good to me

One comment, maybe we can make the email implementation beans `@Named` in case users want to use more than one in the same application.

**Comment by sdelamo** [comment]
> One comment, maybe we can make the email implementation beans `@Named` in case users want to use more than one in the same application.

Good idea. I have annotated them with `@Named` and added a test.

File: email-template/src/main/java/io/micronaut/email/template/DefaultEmailTemplateSender.java [hunk 7]
```diff
@@ -77,7 +78,10 @@ public void send(@NonNull @NotNull @Valid Sender sender,
     }
 
     @NonNull
-    private Optional<String> content(@NonNull ModelAndView<T> modelAndView) {
+    private Optional<String> content(@Nullable ModelAndView<T> modelAndView) {
+        if (modelAndView == null) {
+            return Optional.empty();
+        }
         if (modelAndView.getView().isPresent()) {
             String viewName = modelAndView.getView().get();
```
Comment (ilopmar) [inline email-template/src/main/java/io/micronaut/email/template/DefaultEmailTemplateSender.java]: Maybe you can group both `ifs`: `if (modelAndView != null && modelAndView.getView().isPresent())`

File: gradle.properties [hunk 8]
```diff
@@ -4,13 +4,14 @@ micronautVersion=3.1.4
...
-postmarkVersion=
 title=Micronaut Email
 projectDesc=Integration with Transaction Email Providers
 projectUrl=https://micronaut.io
 githubSlug=micronaut-projects/micronaut-email
 developers=Graeme Rocher
-
+sendgridVersion=4.8.0
+postmarkVersion=1.7.5
+mailjetVersion=5.2.0
```
Comment (ilopmar) [inline gradle.properties:14]: Add them to `bomProperties`

File: email/src/main/java/io/micronaut/email/TransactionalEmail.java [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Representation of a transactional email.
+ * @author Sergio del Amo
+ * @since 1.0.0
+ */
+@Introspected
+public class TransactionalEmail {
```
Comment (musketyr) [inline email/src/main/java/io/micronaut/email/TransactionalEmail.java:35]: I would simply call the class `Email`. I know this is somewhat industry standard but I never understood what transactional about these emails.

File: email/src/main/java/io/micronaut/email/Sender.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+
+/**
+ * @author Sergio del Amo
+ * @since 1.0.0
+ */
+@Introspected
+public class Sender {
```
Comment (musketyr) [inline email/src/main/java/io/micronaut/email/Sender.java:30]: I would inline this class into the email one.

File: email/src/main/java/io/micronaut/email/Recipient.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Email Recipient.
+ * @author Sergio del Amo
+ * @since 1.0.0
+ */
+@Introspected
+public class Recipient {
```
Comment (musketyr) [inline email/src/main/java/io/micronaut/email/Recipient.java:33]: I would inline this class into the email one

File: email/src/main/java/io/micronaut/email/EmailCourier.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+
+/**
+ * Defines a functional interface to send transactional emails.
+ * @author Sergio del Amo
+ * @since 1.0.0
+ */
+@FunctionalInterface
+public interface EmailCourier {
```
Comment (musketyr) [inline email/src/main/java/io/micronaut/email/EmailCourier.java:28]: `EmailSender` sound more natural to me. `Courier` sound more like something physically carrying the emails.

File: email-template/src/main/java/io/micronaut/email/template/EmailTemplateCourier.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Contract for sending an Email template.
+ *
+ * @author Sergio del Amo
+ * @since 1.0
+ * @param <T> The model type
+ */
+@FunctionalInterface
+public interface EmailTemplateCourier<T> {
```
Comment (musketyr) [inline email-template/src/main/java/io/micronaut/email/template/EmailTemplateCourier.java:36]: I would decouple the sending and rendering part completely as there might be multiple implementation of the renderers and multiple implementation of email senders. there could be just a simple interface 

```
interface EmailRenderer {
    Email render(Email templatedEmail, Object model)
}
```

which can be later used as

```
sender.send(renderer.render(email))
```

**Review by musketyr — COMMENTED** [review]
this is a good start. it's missing two important parts

- [ ] support for attachments
- [ ] returning some status from the send method (boolean, enum)

see the individual comments

if you plan to support SES then feel free to take anything from these two projects (but please, support both version of SDK).

https://github.com/agorapulse/micronaut-aws-sdk/tree/master/subprojects/micronaut-amazon-awssdk-ses
https://github.com/agorapulse/micronaut-aws-sdk/tree/master/subprojects/micronaut-aws-sdk-ses

**Comment by morki** [comment]
Nice work, I just could not find way to use `EmailTemplateSender` and use attachments, am I missing something?

**Comment by sdelamo** [comment]
@morki

> I just could not find way to use `EmailTemplateSender` and use attachments, am I missing something?

You are right. It is still something pending. I may add `attachments` to `EmailHeader`but the name of that class looks weird.  I think we should rename it. Not sure if I should rename to:

- `EmailHeader` to `Email` 
- `Email` to `CompleteEmail` 

- `EmailHeader` to `EmailSkeleton`
- `Email` to `Email`

Any naming suggestions, ideas or other feedback is welcome. 

We need to define a POJO as a return type.

**Comment by musketyr** [comment]
> @morki
> 
> > I just could not find way to use `EmailTemplateSender` and use attachments, am I missing something?
> 
> You are right. It is still something pending. I may add `attachments` to `EmailHeader`but the name of that class looks weird. I think we should rename it. Not sure if I should rename to:
> 
> * `EmailHeader` to `Email`
> * `Email` to `CompleteEmail`
> * `EmailHeader` to `EmailSkeleton`
> * `Email` to `Email`
> 
> Any naming suggestions, ideas or other feedback is welcome.
> 
> We need to define a POJO as a return type.

you still have an option to do it as I suggested earlier 😄 

`EmailTemplateSender` would take and  `Email` and render its body and text using the particular internal renderer and send the result.

File: email/src/main/java/io/micronaut/email/Email.java [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+            builder = builder == null ? EmailHeader.builder().bcc(bcc) : builder.bcc(bcc);
+            return this;
+        }
+
+        /**
+         *
+         * @param trackLinks Whether email links should be tracked
+         * @return Email Header Builder
+         */
+        public Builder trackLinks(TrackLinks trackLinks) {
```
Comment (musketyr) [inline email/src/main/java/io/micronaut/email/Email.java:367]: sometimes it's nice to have shortucts for the particular enums e.g. also provide

```
builder.trackLinksInHtml()
builder.trackLinksInHtmlAndText()
builder.trackLinksInText()
```

File: email/src/main/java/io/micronaut/email/EmailSender.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import io.micronaut.core.naming.Named;
+import javax.validation.Valid;
+import javax.validation.constraints.NotNull;
+
+/**
+ * Defines a functional interface to send transactional emails.
+ * @author Sergio del Amo
+ * @since 1.0.0
+ */
+public interface EmailSender extends Named {
```
Comment (musketyr) [inline email/src/main/java/io/micronaut/email/EmailSender.java:28]: add a feature flags so the users know which feature is supported so they can adapt their code

```
default boolean isTrackingLinksSupported() { return false; }
default boolean isSendingAttachmentsSupported() { return false; }
```

File: email/src/main/java/io/micronaut/email/Attachment.java [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+            return this;
+        }
+
+        /**
+         *
+         * @param content file content
+         * @return Attachment's builder
+         */
+        @NonNull
+        public Builder content(@NonNull byte[] content) {
```
Comment (musketyr) [inline email/src/main/java/io/micronaut/email/Attachment.java:148]: it is convenient to add more options to specify content - at least `InputStream` and `File`. there is plenty of responses on SFO how to convert `File` or `InputStream` to `byte[]` but why forcing the developers to Google stuff when you can provide them a solution right ahead.

File: email/src/main/java/io/micronaut/email/EmailSender.java [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+     * @param email Email
+     */
+    void send(@NonNull @NotNull @Valid Email email);
```
Comment (musketyr) [inline email/src/main/java/io/micronaut/email/EmailSender.java:33]: it's very convenient to also provide a builder variant

```
    default void send(Consumer<Email.Builder> email) {
        Email.Builder builder = Email.builder();
        email.accept(builder)
        send(builder.build());
    }
```

File: email/src/main/java/io/micronaut/email/Email.java [hunk 18]
```diff
@@ -0,0 +1,433 @@
...
+            return this;
+        }
+
+        /**
+         *
+         * @param attachment Email's attachment
+         * @return The Transactional Email Builder
+         */
+        @NonNull
+        public Builder attachment(@NonNull Attachment attachment) {
```
Comment (musketyr) [inline email/src/main/java/io/micronaut/email/Email.java:410]: it's very convenient to also add builder version

```
public Builder attachment(Consumer<Attachment.Builder> attachment) {
    Attachment.Builder builder = Attachment.builder();
    attachment.accept(builder);
    return attachment(builder.build());
}
```

**Comment by musketyr** [comment]
see some usability recommendations. if Groovy is still one of the languages it might be nice if you also provide `Closure` support for the `Consumer<X.Builder>` methods 

see https://medium.com/p/99bd118538b4

**Review by musketyr — CHANGES_REQUESTED** [review]

**Comment by morki** [comment]
> `EmailTemplateSender` would take and `Email` and render its body and text using the particular internal renderer and send the result.

What if there will be only one `Email` class, which can contain arbitrary `context: Map<String, Any>` and `EmailTemplateSender` was just a one of registered `EmailMapper`s of that `Email` class.

This can open door to some situations (we are dealing now with similar home grown solution):
- users can make customization with only creating new simple class
- we for example use this to populate subject from parsed HTML template
- internationalization of all parts like subject, text, html
- adding FROM in one centralized place

**Comment by sdelamo** [comment]
> see some usability recommendations. if Groovy is still one of the languages it might be nice if you also provide `Closure` support for the `Consumer<X.Builder>` methods
> 
> see https://medium.com/p/99bd118538b4

I don't really understand your remark about Groovy being still one of the languages. Yes, Groovy is supported in the framework as much as we can. I don't think any other framework apart from Grails supports Groovy more than Micronaut.  

I will be happy to have a `email-groovy` module with enhancements to the API to have a better experience for Groovy users. PRs are welcomed against this branch. 

Thanks for your other suggestions. I have added them all.

**Comment by sdelamo** [comment]
@musketyr @morki I have changed the APIs and I think it is now easy for the users. I have created an Email builder in different packages with the same API. They have always a Email builder api for both plain emails or emails with templates only with the methods applicable for each of these use cases.  Let me know what you think. 

> users can make customization with only creating new simple class

`io.micronaut.email.Email` is a public class  which users can extend. 

> we for example use this to populate subject from parsed HTML template

> internationalization of all parts like subject, text, html

We have to improve the APIs for internationalization but I am not sure this should be in the email module.  I expect users to localize their subject and models and then call `EmailSender::send` with localized strings. 

> adding FROM in one centralized place

I think this is possible with the current implementation. 

I expect people to have a singleton in their applications where they inject a configuration with the common from, an `EmailSender` and they will do logic with the response type. I am not sure we need to implement this logic in this module. It seems to me it will vary from application to application.

**Comment by morki** [comment]
Thank you @sdelamo for your cool work. Just one last thing: FROM is now `@NonNull`, but using single FROM centralized in configuration is common practice (maybe only for us). What do you think?

**Comment by sdelamo** [comment]
@morki 

> Thank you @sdelamo for your cool work. Just one last thing: FROM is now `@NonNull`, but using single FROM centralized in configuration is common practice (maybe only for us). What do you think?

I think from needs to be required. However, I see your use case and others (always bcc an address, prefix subject). Thus, I have added a feature `EmailDecorator` to achieve this. Generate documentation in this branch  `./gradlew docs`, check `Decorators` section and let me know what you think.

**Comment by morki** [comment]
Really nice @sdelamo, this is exactly what we needed :+1:  Maybe remove `@NonNull` from `from` in both `Email` classes. Again, thank you very much for your work.

**Comment by sdelamo** [comment]
> Really nice @sdelamo, this is exactly what we needed 👍 Maybe remove `@NonNull` from `from` in both `Email` classes. Again, thank you very much for your work.

I cannot remove `@NonNull`. At least not in `io.micronaut.email.Email`. I am relying on that constraint to ensure `TransactionEmailSender`is only invoked with valid emails. See `EmailValidatedSpec`

**Comment by sdelamo** [comment]
@jameskleeh Could you check this PR. I would like to get it release for Micronaut 3.3. ?

File: gradle.properties (hunk 8 above)
Comment (timyates) [reply gradle.properties:15]: There's been some new updates of libraries:

- a [4.8.2 version](https://search.maven.org/artifact/com.sendgrid/sendgrid-java) of sendgrid-java now 
- a [1.8.0 version](https://search.maven.org/artifact/com.wildbit.java/postmark) of postmark
- and a [5.2.0 version](https://search.maven.org/artifact/org.apache.poi/poi-ooxml) of poi-ooxml

Do we want to use the latest?

File: src/main/docs/guide/integrations/javamail.adoc [hunk 19]
```diff
@@ -0,0 +1,5 @@
+To use https://javaee.github.io/javamail/[JavaMail], add the following dependency to your application.
+
+dependency:micronaut-javamail[groupId="io.micronaut.email"]
```
Comment (timyates) [inline src/main/docs/guide/integrations/javamail.adoc:3]: Should be `micronaut-email-javamail` I believe 🤔

**Review by timyates — CHANGES_REQUESTED** [review]

**Comment by timyates** [comment]
We should change

https://github.com/micronaut-projects/micronaut-email/blob/075214a299f3115a1030fccd51af23b2175b5645/.github/workflows/gradle.yml#L22

to `[8, 11, 17]` as well instead of `[8, 11, 16]`

File: gradle.properties (hunk 8 above)
Comment (graemerocher) [reply gradle.properties:15]: Given the discussion in the last call this project needs a BOM and doesn't need to add them to `bomProperties`

**Review by timyates — APPROVED** [review]

**Comment by sdelamo** [comment]
@jameskleeh I have included your PR, can you please review (approve / request changes) again?

**Review by graemerocher — APPROVED** [review]

**Pull request merged**